llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp

```diff
@@ -13760,6 +13760,26 @@ Value *BoUpSLP::vectorizeTree(TreeEntry *E, bool PostponedPHIs) {
           LLVM_DEBUG(dbgs() << "SLP: Diamond merged for " << *VL0 << ".\n");
           return E->VectorizedValue;
         }
+        if (isa<FixedVectorType>(ScalarTy)) {
+          assert(SLPReVec && "FixedVectorType is not expected.");
+          // CreateMaskedGather expects VecTy and VecPtr have same size. We need
+          // to expand VecPtr if ScalarTy is a vector type.
+          unsigned ScalarTyNumElements =
+              cast<FixedVectorType>(ScalarTy)->getNumElements();
+          unsigned VecTyNumElements =
+              cast<FixedVectorType>(VecTy)->getNumElements();
+          SmallVector<Constant *> Indices(VecTyNumElements);
+          transform(seq(VecTyNumElements), Indices.begin(), [=](unsigned I) {
+            return Builder.getInt64(I % ScalarTyNumElements);
+          });
```

Comment:
Are we rereading the same data here several times? Can it be replaced with subvector insert instead?

Reply:
No, I mean read the data once and then just copy it several times in IR

Reply:
I don't think we can. Each time the ScalarTyNumElements is different.

Reply:
Why? It may vary between calls, but it is not modified here. Instead of re-reading the same data several times, better just read it once and then insert subvector(s)

Reply:
The data is read several time, but is not read in sequence. If we use CreateInsertVector to build the result, we still need CreateShuffleVector to permute.

Reply:
Could you provide a test to prove this? The existing test has pattern 0,1,0,1, which can be replaced by just subvector insert.

Reply:
This case is trying to vectorize the following instructions
```
  %getelementptr0 = getelementptr i8, ptr null, i64 64036
  %getelementptr1 = getelementptr i8, ptr null, i64 64064
  %0 = load <2 x i32>, ptr %getelementptr0, align 4
  %1 = load <2 x i32>, ptr %getelementptr1, align 8
```
So the actual load address is `[64036, 64036 + i32, 64064, 64064 + i32]`.

The `VecPtr` is `<4 x ptr> getelementptr (i32, <4 x ptr> <ptr inttoptr (i64 64036 to ptr), ptr inttoptr (i64 64036 to ptr), ptr inttoptr (i64 64064 to ptr), ptr inttoptr (i64 64064 to ptr)>, <4 x i64> <i64 0, i64 1, i64 0, i64 1>)`.
From the definition, I though the meaning for `0, 1, 0, 1` is

```
[0] = 64036 + 0 * sizeof(i32)
[1] = 64036 + 1 * sizeof(i32)
[2] = 64064 + 0 * sizeof(i32)
[3] = 64064 + 1 * sizeof(i32)
```

reference: https://llvm.org/docs/LangRef.html#vector-of-pointers

Reply:
Masked gather is worse than 2 consecutive loads + insert vector. Why do we want to represent as masked gather?

Reply:
Because this PR aims to make the algorithm function properly instead of crashing.
Or we do not vectorize the instructions if ScalarTy is FixedVectorType?

Reply:
Why not represent it as 2 loads + insert?

Reply:
Also, do you have the cost model for this change?

Reply:
> Also, do you have the cost model for this change?

GEPs are calculated by `TTI.getGEPCost` in `getGEPCosts`.
`TTI->getGatherScatterOpCost` correctly calculates the cost if ScalarTy is FixedVectorType.

Reply:
> > Why not represent it as 2 loads + insert?
> 
> This test has `-slp-threshold=-100`. Basically, scalar version (loads + insert) cost is smaller than vector version (masked gather) cost.

By `scalar` you mean small vector `<2 x >`?

Reply:
Yes.